File: npm/src/capture.ts
import { InvalidURLError } from "./errors";
import { Session, normalizeUrl } from "./engine";
import {
  CaptureResult,
  CapturedPage,
  CrawlResult,
  DEFAULT_VIEWPORTS,
  Viewport,
} from "./models";

export interface CaptureOptions {
  /** "all" = desktop + tablet + mobile; "desktop" = 1440 only. */
  mode?: "all" | "desktop";
  /** Explicit viewports (overrides `mode`). */
  viewports?: Viewport[];
  /** Prime hover to surface sprite/hover-only artwork (default true). */
  extractHover?: boolean;
  imageFormat?: "jpeg" | "png";
  /** JPEG quality 0–100 (ignored for PNG). Default 62. */
  quality?: number;
  /** Navigation timeout in seconds. Default 15. */
  timeout?: number;
  /** Custom Chromium/Chrome binary; defaults to Puppeteer's bundled browser. */
  executablePath?: string;
  /** Run headless (default true). */
  headless?: boolean;
}

export interface CrawlOptions extends CaptureOptions {
  /** Maximum number of pages to capture. Default 5. */
  maxPages?: number;
  /** URLs to seed the queue with first. */
  customUrls?: string[];
}

function resolveViewports(mode: string, viewports?: Viewport[]): Viewport[] {
  if (viewports && viewports.length) return viewports;
  if (mode === "desktop") return DEFAULT_VIEWPORTS.filter((v) => v.key === "desktop");
  return DEFAULT_VIEWPORTS;
}

function slugForPage(url: string, index: number): string {
  try {
    const p = new URL(url).pathname.replace(/^\/+|\/+$/g, "");
    const s = p.replace(/[^a-z0-9]+/gi, "-").replace(/^-+|-+$/g, "");
    return s || (index === 0 ? "home" : `page-${index}`);
  } catch {
    return index === 0 ? "home" : `page-${index}`;
  }
}

/** Capture a single page's visual blueprint. */
export async function capture(
  url: string,
  options: CaptureOptions = {},
): Promise<CaptureResult> {
  const target = normalizeUrl(url);
  if (!target) throw new InvalidURLError(`Not a valid website URL: ${url}`);

  const {
    mode = "all",
    viewports,
    extractHover = true,
    imageFormat = "jpeg",
    quality = 62,
    timeout = 15,
    executablePath,
    headless = true,
  } = options;

  const picked = resolveViewports(mode, viewports);
  const session = new Session({ headless, executablePath });
  await session.start();
  try {
    await session.goto(target, timeout * 1000);

    const result = new CaptureResult({
      url: target,
      title: await session.title(target),
      capturedAt: new Date().toISOString(),
      imageFormat,
    });

    await session.autoScroll();

    const rest = await session.extract();
    result.colors = rest.colors;
    result.fonts = rest.fonts;

    const images = [...rest.images];
    if (extractHover) {
      await session.primeHover();
      const hovered = await session.extract();
      await session.resetHover();
      for (const img of hovered.images) {
        if (!images.includes(img)) images.push(img);
      }
    }
    images.sort(
      (a, b) => (a.startsWith("data:") ? 0 : 1) - (b.startsWith("data:") ? 0 : 1),
    );
    result.images = images.slice(0, 40);

    for (const vp of picked) {
      result.screenshots[vp.key] = await session.screenshot(vp, imageFormat, quality);
    }
    return result;
  } finally {
    await session.close();
  }
}

/** Crawl a domain (breadth-first, same-origin), capturing each page. */
export async function crawl(
  url: string,
  options: CrawlOptions = {},
): Promise<CrawlResult> {
  const start = normalizeUrl(url);
  if (!start) throw new InvalidURLError(`Not a valid website URL: ${url}`);

  const {
    maxPages = 5,
    customUrls = [],
    mode = "desktop",
    viewports,
    imageFormat = "jpeg",
    quality = 62,
    timeout = 15,
    executablePath,
    headless = true,
  } = options;

  const origin = new URL(start).origin;
  const picked = resolveViewports(mode, viewports);
  const session = new Session({ headless, executablePath });
  await session.start();

  const crawlRes = new CrawlResult({ domain: new URL(start).hostname });
  const seeds = customUrls
    .map((u) => normalizeUrl(u))
    .filter((u): u is string => Boolean(u));
  const queue: string[] = [...seeds, start];
  const seen = new Set<string>();

  try {
    while (queue.length && crawlRes.pages.length < maxPages) {
      const nextRaw = queue.shift()!;
      const next = nextRaw.split("#")[0];
      if (seen.has(next)) continue;
      seen.add(next);

      try {
        await session.goto(next, timeout * 1000);
      } catch {
        continue; // skip pages that don't load
      }

      await session.autoScroll();
      const rest = await session.extract();
      const page = new CapturedPage({
        url: next,
        slug: slugForPage(next, crawlRes.pages.length),
        title: await session.title(next),
        colors: rest.colors,
        fonts: rest.fonts,
        imageFormat,
      });
      for (const vp of picked) {
        page.screenshots[vp.key] = await session.screenshot(vp, imageFormat, quality);
      }
      crawlRes.pages.push(page);
      crawlRes.sitemap.push(next);

      const links: string[] = await session.page.evaluate(() =>
        Array.from(document.querySelectorAll("a[href]")).map(
          (a) => (a as HTMLAnchorElement).href,
        ),
      );
      for (const l of links) {
        try {
          const lu = new URL(l);
          if (lu.origin !== origin) continue;
          const clean = lu.origin + lu.pathname;
          if (!seen.has(clean)) queue.push(clean);
        } catch {
          /* ignore bad hrefs */
        }
      }
    }
    return crawlRes;
  } finally {
    await session.close();
  }
}

File: python/src/webforge/capture.py
"""Public capture orchestration: :func:`capture` and :func:`crawl`.

These are re-exported at the top level, so the intended usage is simply::

    import webforge

    result = webforge.capture("example.com")
    result.to_zip("example.zip")
"""

from __future__ import annotations

from datetime import datetime, timezone
from urllib.parse import urlparse

from ._engine import _Session, normalize_url
from .errors import InvalidURLError
from .models import DEFAULT_VIEWPORTS, CapturedPage, CaptureResult, CrawlResult, Viewport

__all__ = ["capture", "crawl"]


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def _resolve_viewports(mode: str, viewports) -> tuple[Viewport, ...]:
    """Turn a ``mode`` / explicit ``viewports`` argument into a viewport tuple."""
    if viewports is not None:
        resolved = tuple(viewports)
        if not resolved:
            raise ValueError("viewports must not be empty")
        return resolved
    if mode == "desktop":
        return tuple(v for v in DEFAULT_VIEWPORTS if v.key == "desktop")
    if mode == "all":
        return DEFAULT_VIEWPORTS
    raise ValueError(f"mode must be 'all' or 'desktop', got {mode!r}")


def _slug_from(url: str) -> str:
    path = urlparse(url).path
    return path or "/"


def capture(
    url: str,
    *,
    mode: str = "all",
    viewports=None,
    extract_hover: bool = True,
    image_format: str = "jpeg",
    quality: int = 62,
    timeout: float = 15.0,
    executable_path: str | None = None,
    headless: bool = True,
) -> CaptureResult:
    """Capture a single page's visual blueprint.

    Args:
        url: Target URL. A missing scheme defaults to ``https://``.
        mode: ``"all"`` captures desktop + tablet + mobile; ``"desktop"`` only
            the 1440-wide shot. Ignored when ``viewports`` is given.
        viewports: Optional explicit iterable of :class:`~webforge.Viewport`.
        extract_hover: Also prime hover to surface sprite/hover-only artwork.
        image_format: ``"jpeg"`` (default, smaller) or ``"png"``.
        quality: JPEG quality 0-100 (ignored for PNG).
        timeout: Seconds to wait for navigation before capturing current state.
        executable_path: Custom Chromium binary; defaults to Playwright's.
        headless: Run the browser headless (default ``True``).

    Returns:
        A :class:`~webforge.CaptureResult` with screenshots, colours, fonts and images.

    Raises:
        InvalidURLError: If ``url`` is empty or not http(s).
        BrowserNotInstalledError: If Playwright/Chromium is unavailable.
    """
    target = normalize_url(url)
    if target is None:
        raise InvalidURLError(f"Not a valid website URL: {url!r}")

    picked = _resolve_viewports(mode, viewports)

    with _Session(executable_path=executable_path, headless=headless) as session:
        session.goto(target, timeout=timeout)
        result = CaptureResult(
            url=target,
            title=session.title(target),
            captured_at=_now_iso(),
            image_format=image_format,
        )

        session.auto_scroll()

        # Tokens at rest give the default artwork, colours and fonts.
        rest = session.extract_tokens()
        result.colors = rest["colors"]
        result.fonts = rest["fonts"]

        images = list(rest["images"])
        if extract_hover:
            # Hovering can swap art in (and replace resting art), so we prime
            # hover purely for the extra images, then reset before screenshots.
            session.prime_hover()
            hovered = session.extract_tokens()
            session.reset_hover()
            for img in hovered["images"]:
                if img not in images:
                    images.append(img)

        # data-URIs first, capped at 40 -- matches the route's ordering.
        images.sort(key=lambda s: 0 if s.startswith("data:") else 1)
        result.images = images[:40]

        for vp in picked:
            result.screenshots[vp.key] = session.screenshot(vp, image_format, quality)

    return result


def crawl(
    url: str,
    *,
    max_pages: int = 5,
    custom_urls=None,
    mode: str = "desktop",
    viewports=None,
    image_format: str = "jpeg",
    quality: int = 62,
    timeout: float = 15.0,
    executable_path: str | None = None,
    headless: bool = True,
) -> CrawlResult:
    """Crawl a domain, capturing each page it reaches (breadth-first).

    Only same-origin http(s) links are followed, matching the extension's
    crawler. ``custom_urls`` are visited first, then the start URL, then links
    discovered along the way, until ``max_pages`` pages are captured.

    Args:
        url: The start URL (its origin bounds the crawl).
        max_pages: Maximum number of pages to capture.
        custom_urls: Optional iterable of URLs to seed the queue with first.
        mode: ``"desktop"`` (default) or ``"all"`` viewports per page.
        viewports: Optional explicit viewports (overrides ``mode``).
        image_format, quality, timeout, executable_path, headless: As :func:`capture`.

    Returns:
        A :class:`~webforge.CrawlResult` with every captured page and the sitemap.

    Raises:
        InvalidURLError: If ``url`` is empty or not http(s).
    """
    start = normalize_url(url)
    if start is None:
        raise InvalidURLError(f"Not a valid website URL: {url!r}")

    picked = _resolve_viewports(mode, viewports)
    origin = urlparse(start).scheme + "://" + urlparse(start).netloc

    seeded = list(custom_urls or [])
    if start not in seeded:
        seeded.append(start)

    queue: list[str] = list(seeded)
    visited: set[str] = set()
    result = CrawlResult(domain=urlparse(start).netloc)

    with _Session(executable_path=executable_path, headless=headless) as session:
        while queue and len(result.pages) < max_pages:
            current = queue.pop(0)
            if current in visited:
                continue
            visited.add(current)

            session.goto(current, timeout=timeout)
            session.auto_scroll()
            tokens = session.extract_tokens()

            page = CapturedPage(
                url=current,
                slug=_slug_from(current),
                title=session.title(current),
                colors=tokens["colors"],
                fonts=tokens["fonts"],
                image_format=image_format,
            )
            for vp in picked:
                page.screenshots[vp.key] = session.screenshot(vp, image_format, quality)
            result.pages.append(page)

            # Discover more same-origin links to keep crawling.
            for link in session.same_origin_links(origin):
                if link not in visited and link not in queue:
                    queue.append(link)

    result.sitemap = sorted(visited)
    return result
